### ISSUE TITLE
Order of execution of listeners

### ISSUE THREAD

**Issue by kanthicn1** [google/ExoPlayer#4613]
Hi,

I am using ExoPlayer 2.8.2 release and trying to implement a feature where I want to use the bandwidth measurement from the latest video segment download to influence the quality of the next segment. This is for Dash videos.
I have extended trackSelection to `CustomTrackSelection` object. But, the problem on several occasions, by the time the `DefaultDashChunkSource` calls `updateSelectedTrack` function, the previous download end event hasn't been processed by my `listener` and I haven't had a chance to register the latest throughput sample. 
I guess what's happening is that the `onLoadCompleted` event is processed by `ChunkSampleStream` and it proceeds with the next download start _before_ my listener is invoked.
Is there any way to get around this? Is there any way I can "wait" before downloading the next segment if my listener hasn't registered the download end event?

Thanks for your attention and help! I appreciate any pointers.

**Comment by erdemguven** [comment]
Why don't you use AdaptiveTrackSelection? It already selects tracks according to bandwidth measurement.

**Comment by kanthicn1** [comment]
Fair enough @erdemguven . Sorry for not writing the complete problem. I don't want to use `AdaptiveTrackSelection` because I want to use my own track selection logic which doesn't quite agree with what `AdaptiveTrackSelection` is doing. Also, the `DefaultBandwidthMeter`, which is used by `AdaptiveTrackSelection`,  gives a  "SlidingPercentile" estimate of bandwidth while I want the last instantaneous value of the bandwidth measurement. 
One approach that might actually work for my case is to use a different bandwidth meter instead of relying on throughput measurements from listeners.

(a) implement `CustomBandwidthMeter` and pass that to the `buildDataSourceFactory` and `customTrackSelection`. -- this is so that my `CustomBandwidthMeter` can expose the last instantaneous measurement of bandwidth sample.
(b) Query the `CustomBandwidthMeter` for the latest throughput sample from `CustomTrackSelection` and use it in my track selection logic.

My question is: in the above logic, is there a chance that I might not get the latest throughput sample due to race conditions in event processing? i.e., Is the `DataSourceFactory` guaranteed to update the bandwidth sample before `CustomTrackSelection` is called for picking the quality for the next segment to be loaded?

**Comment by erdemguven** [comment]
@ojw28 could you look at this?

**Comment by ojw28** [comment]
> My question is: in the above logic, is there a chance that I might not get the latest throughput sample due to race conditions in event processing? i.e., Is the DataSourceFactory guaranteed to update the bandwidth sample before CustomTrackSelection is called for picking the quality for the next segment to be loaded?

There is a guarantee, which is that the `TransferListener.onTransferEnd` call corresponding to a chunk will always be called before `TrackSelection.updateSelectedTrack` is called to pick the quality for the next chunk. Note that `DefaultBandwidthMeter` updates its bitrate estimate in `onTransferEnd`, and `AdaptiveTrackSelection` gets the bitrate estimate in `TrackSelection.updateSelectedTrack`. Hence it's guaranteed that the updated bitrate estimate is used.

So in short, as long as you follow the same model in your custom components (albeit changing the actual logic for estimating the bitrate and/or choosing the track), then you'll get the same guarantee, which I think is what you want.

**Comment by kanthicn1** [comment]
Perfect. Thanks @ojw28 and @erdemguven.
